cupyx/scipy/ndimage/_filters_generic.py

```diff
@@ -0,0 +1,533 @@
+import re
+import types
+
+import cupy
+
+from cupyx.scipy.ndimage import _filters_core
+
+
+def _get_sub_kernel(f, filter_size=0, in_dtype=cupy.dtype(float)):
+    """
+    Takes the "function" given to generic_filter and returns the "sub-kernel"
+    that will be called, one of RawKernel, ReductionKernel, or FusedKernel.
+
+    This supports:
+     * cupy.RawKernel
+       no checks are possible
+     * cupy.ReductionKernel
+       checks that there is a single input and output
+     * cupy.core._fusion_kernel.FusedKernel
+       checks that there is a single input and output and some of the
+       properties of those arguments
+     * cupy.core.fusion.Fusion and cupy.core.new_fusion.Fusion
+       retrieves underlying kernel and sends it back through this function
+     * any callable
+       attempts to fuse it then and sends it back through this function
+
+    The filter_size and in_dtype arguments are the size and data type used for
+    function fusions which requires calling the fused function with an array of
+    the appropriate size and data type to create the underlying kernel.
+    """
+    if isinstance(f, cupy.RawKernel):
+        # We will assume that it has the correct API
+        return f
+    elif isinstance(f, cupy.ReductionKernel):
+        if f.nin != 1 or f.nout != 1:
+            raise TypeError('ReductionKernel must have 1 input and output')
+        return f
+    elif isinstance(f, cupy.ElementwiseKernel):
+        # special error message for ElementwiseKernels
+        raise TypeError('only ReductionKernel allowed (not ElementwiseKernel)')
+    elif isinstance(f, cupy.core._fusion_kernel.FusedKernel):
+        out = f._params[f._out_params[0]] if f._return_size == -2 else None
+        if (sum(i >= 0 for i in f._input_index) != 1 or
+                not isinstance(out, cupy.core._fusion_variable._TraceArray) or
+                out.ndim != 0):
+            raise TypeError('fused function take must take 1 array argument '
+                            'and output 1 scalar')
+        return f
+
+    # The remaining ones need additional (recursive) processing
+    elif isinstance(f, cupy.core.new_fusion.Fusion):
+        kernel = _get_new_fused_kernel(f, filter_size, in_dtype)
+    elif isinstance(f, cupy.core.fusion.Fusion):
+        kernel = getattr(f, 'new_fusion', None)
+        if kernel is None:
+            kernel = _get_fused_kernel(f, filter_size, in_dtype)
+    elif callable(f):
+        kernel = cupy.fuse(f)
+    else:
+        raise TypeError('bad function type')
+    return _get_sub_kernel(kernel, filter_size, in_dtype)
+
+
+def _get_fused_kernel(func, filter_size, in_dtype):
+    # This may have to call the fused function so that it generates the
+    # reduction kernel for our datatype
```

Comment:
I think that these functions can be joined and moved to cupy/core/_fusion* as they are quite generic.

Reply:
Yeah, basically getting the underlying callable for a given set of inputs (preferably with just shapes/dtypes of inputs, but understandable if it has to be the actual inputs themselves) but without actually calling it. It would be very nice if that were available on fused functions.